### DISCUSSION TITLE
DAO/ Raw query operations after receiving body from upstream

### DISCUSSION THREAD

**Discussion by akki005** [Kong/kong#12168] [Help]
We have a use case where we want to operation in following order in the kong `plugin`.

`read the response body received from the upstream->generate tokens->save tokens to the kong db->add it in response header`

1. Was trying this with body_filter phase but i can see that db related operation are not allowed in `body_filter` phase. 
2. tried running raw db query in body_filter phase but was not able to do `connector:get_stored_connection` as it is returning nil.

What are the possible solution for this with kong?

**Comment by StarlightIbuki** [comment]
Please refer to: https://github.com/openresty/lua-nginx-module#cosockets-not-available-everywhere
Even without the limitation, the body filter is not supposed to be used for this purpose as it could be executed multiple times.

If the timing for saving the token into the DB does not matter, you could start a timer to do the job, and the timer should be started only once for a request (you can ensure this with a status flag in `ngx.ctx`). If the timing matters, you could use a lock to block the request.

Also, it could be tricky to set a response header after reading the whole response body, as the body is streaming served and the header is already sent.
